ci/scripts/python_test.sh

```diff
@@ -35,6 +35,9 @@ export PYTHONDEVMODE=1
 # Enable memory debug checks.
 export ARROW_DEBUG_MEMORY_POOL=trap
 
+# Retain directories only for tests with outcome error or failed
```

Comment:
Maybe we can try to uninstall pytest and install the older version on this file.

Reply:
That would be a good option but I would first try to somehow pass `tmp_path_retention_policy=failed` so that we do not have to pin the version and keep using the latest released pytest. I just need to figure out how to do that with `PYTEST_ADDOPTS` 😊

Reply:
Actually, pinning the version until the fix for the issue is in a released version of pytest would be a better option (I just looked at the connected pytest issue and saw there was some new information).

Reply:
In the meantime, pytest 7.3.1 has been released: https://github.com/pytest-dev/pytest/releases/tag/7.3.1

Reply:
I see AppVeyor and MacOS build are working fine with pytest 7.3.1 (on one of the PRs updated today):

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/46801690
https://github.com/apache/arrow/actions/runs/4717563273/jobs/8366298024?pr=34730

Windows builds seem to take the same time as they did before the 7.3. release:
- 9 Apr https://github.com/ursacomputing/crossbow/actions/runs/4649634532/jobs/8228109815
- yesterday https://github.com/ursacomputing/crossbow/actions/runs/4712356112/jobs/8357330588

I think this PR and the linked issue can be closed. Will do it later today if you agree.